File: src/creational/abstractfactory/simpleimpl/SimpleFactory.java
package creational.abstractfactory.simpleimpl;

import creational.abstractfactory.interfaces.Button;
import creational.abstractfactory.interfaces.WidgetFactory;
import creational.abstractfactory.interfaces.Window;

public class SimpleFactory implements WidgetFactory {
    private static SimpleFactory ourInstance = new SimpleFactory();

    public static SimpleFactory getInstance() {
        return ourInstance;
    }

    private SimpleFactory() {
        System.out.println("SimpleFactory");
    }

    @Override
    public Window createWindow() {
        return new SimpleWindow();
    }

    @Override
    public Button createButton() {
        return new SimpleButton();
    }
}

File: src/creational/abstractfactory/simpleimpl/SimpleFactory.scala
package creational.abstractfactory.simpleimpl

import creational.abstractfactory.interfaces.{Button, TextField, WidgetFactory, Window}

object SimpleFactory extends WidgetFactory {
  override def createWindow(): Window = new SimpleWindow

  override def createButton(): Button = new SimpleButton

  override def createTextField(): TextField = new SimpleTextField
}
